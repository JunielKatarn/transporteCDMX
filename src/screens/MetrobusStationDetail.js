import React, { PureComponent } from 'react'
import { View, Text, FlatList } from 'react-native'
import moment from 'moment'
import { sharedStyles } from '../styles'
import Loader from '../components/Loader'
import SimplePlaceholder from '../components/SimplePlaceholder'
import { getMetrobusStationSchedule } from '../api/endpoints'
import { HTTP_SUCCESS } from '../api/request'
import Metrobus from '../models/Metrobus'
import MetrobusSchedule from '../components/MetrobusSchedule'
import ListHeader from '../components/ListHeader'
import { keyExtractor } from '../utils/list'
import { translate } from '../i18n'

class MetrobusStationDetail extends PureComponent {
  state = {
    metrobusSchedule: null,
    lastUpdate: null,
    isLoading: true,
    refreshing: false,
    error: false,
  }

  componentDidMount() {
    const { station, lineColor } = this.props.route.params
    this.props.navigation.setOptions({ title: station.name, headerTintColor: lineColor })
    this._requestStationSchedule(station.id)
  }

  _setError = () => {
    this.setState({
      metrobusSchedule: null,
      isLoading: false,
      error: true,
      refreshing: false,
      lastUpdate: 'Error',
    })
  }

  _metrobusResponseSuccess = metrobusResponse => {
    if (metrobusResponse?.data) {
      const metrobusSchedule = metrobusResponse.data.response?.metrobuses.map(metrobus => {
        return new Metrobus(metrobus)
      })
      this.setState({
        metrobusSchedule,
        error: false,
        isLoading: false,
        refreshing: false,
        lastUpdate: moment().format('hh:mm:ss A '),
      })
    } else {
      this._setError()
    }
  }

  _handleMetrobusScheduleResponse = metrobusResponse => {
    if (metrobusResponse?.data?.code === HTTP_SUCCESS) {
      this._metrobusResponseSuccess(metrobusResponse)
    } else {
      this._setError()
    }
  }

  _requestStationSchedule = async stationId => {
    try {
      this._handleMetrobusScheduleResponse(await getMetrobusStationSchedule(stationId))
    } catch (e) {
      console.log(e)
      this._setError()
    }
  }

  _renderErrorPlaceholder = () => (
    <SimplePlaceholder
      imageURL={require('../assets/images/network.png')}
      text={translate('error')}
    />
  )

  _renderItem = ({ item }) => (
    <MetrobusSchedule
      metrobus={item}
      navigation={this.props.navigation}
      lineColor={this.props.route.params.lineColor}
    />
  )

  _renderHeader = () =>
    this.state.lastUpdate && (
      <ListHeader text={`${translate('last_updated')} - ${this.state.lastUpdate}`} />
    )

  _renderEmptyPlaceholder = () => (
    <SimplePlaceholder
      imageURL={require('../assets/images/bus_driver.png')}
      text={translate('no_more_buses')}
      fullHeight
    />
  )

  _renderFooter = () => <View style={sharedStyles.mb3} />

  _pullToRefresh = () => {
    const { station } = this.props.route.params
    this.setState({ refreshing: true })
    this._requestStationSchedule(station.id)
  }

  render() {
    const { lineColor } = this.props.route.params
    if (this.state.isLoading) {
      return <Loader containerStyle={sharedStyles.flex1} color={lineColor} visible />
    }

    if (this.state.error) {
      return this._renderErrorPlaceholder()
    }

    return (
      <FlatList
        data={this.state.metrobusSchedule}
        refreshing={this.state.refreshing}
        onRefresh={this._pullToRefresh}
        renderItem={this._renderItem}
        keyExtractor={keyExtractor}
        ListHeaderComponent={this._renderHeader}
        ListEmptyComponent={this._renderEmptyPlaceholder}
        ListFooterComponent={this._renderFooter}
        style={[sharedStyles.fullRelativeHeight, sharedStyles.p3]}
      />
    )
  }
}

export default MetrobusStationDetail
