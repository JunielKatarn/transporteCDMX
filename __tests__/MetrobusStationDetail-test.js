import React from 'react'
import { shallow } from 'enzyme'
import MetrobusStationDetail from '../src/screens/MetrobusStationDetail'
import metrobusStation from './fixtures/metrobusStation-fixture'
import metrobus from './fixtures/metrobus-fixture'
import navigation from './fixtures/navigation-fixture'

const params = { params: { station: metrobusStation, lineColor: '#042EC6' } }
const metrobusSchedule = [metrobus]

describe('MetrobusStationDetail', () => {
  it('it renders correctly while loading', () => {
    const wrapper = shallow(<MetrobusStationDetail navigation={navigation} route={params} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly WITH Metrobus schedule', () => {
    const wrapper = shallow(<MetrobusStationDetail navigation={navigation} route={params} />)
    wrapper.setState({ isLoading: false, metrobusSchedule })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly WITHOUT Metrobus schedule', () => {
    const wrapper = shallow(<MetrobusStationDetail navigation={navigation} route={params} />)
    wrapper.setState({ isLoading: false, metrobusSchedule: [] })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly WITH error', () => {
    const wrapper = shallow(<MetrobusStationDetail navigation={navigation} route={params} />)
    wrapper.setState({ isLoading: false, error: true })
    expect(wrapper).toMatchSnapshot()
  })
})
