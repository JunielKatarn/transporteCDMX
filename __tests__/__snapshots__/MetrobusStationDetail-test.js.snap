// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetrobusStationDetail it renders correctly WITH Metrobus schedule 1`] = `
<FlatList
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "coordinates": Object {
          "latitude": 19,
          "latitudeDelta": 0.01,
          "longitude": -79,
          "longitudeDelta": 0.01,
        },
        "distance": 2020,
        "eta": "2 min",
        "formattedDate": "2020-01-01",
        "id": 1234567,
        "name": "This a name",
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  onRefresh={[Function]}
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  style={
    Array [
      Object {
        "height": "100%",
      },
      Object {
        "padding": 17,
      },
    ]
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`MetrobusStationDetail it renders correctly WITH error 1`] = `
<Memo(SimplePlaceholder)
  imageURL={
    Object {
      "testUri": "../../../src/assets/images/network.png",
    }
  }
  text="[missing \\"en.error\\" translation]"
/>
`;

exports[`MetrobusStationDetail it renders correctly WITHOUT Metrobus schedule 1`] = `
<FlatList
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  ListHeaderComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  onRefresh={[Function]}
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  style={
    Array [
      Object {
        "height": "100%",
      },
      Object {
        "padding": 17,
      },
    ]
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`MetrobusStationDetail it renders correctly while loading 1`] = `
<Memo(Loader)
  color="#042EC6"
  containerStyle={
    Object {
      "flex": 1,
    }
  }
  visible={true}
/>
`;
